imap_data_access/file_validation.py

```diff
@@ -0,0 +1,226 @@
+"""Methods for managing and validating filenames and filepaths"""
+from __future__ import annotations
+
+import re
+from datetime import datetime
+from pathlib import Path
+
+import imap_data_access
+
+
+class ScienceFilepath:
+    class InvalidScienceFileError(Exception):
+        """Indicates a bad file type"""
+
+        pass
+
+    def __init__(self, filename: str | Path, data_dir: Path | None = None):
+        """Class to store filepath and file management methods for science files.
+
+        If you have an instance of this class, you can be confident you have a valid
+        science file and generate paths in the correct format.
+
+        Current filename convention:
+        <mission>_<instrument>_<datalevel>_<descriptor>_<startdate>_<enddate>
+        _<version>.<extension>
+
+        NOTE: There are no optional parameters anymore. All parameters are required.
+        <mission>: imap
+        <instrument>: idex, swe, swapi, hi-45, ultra-45 and etc.
+        <datalevel> : l1a, l1b, l1, l3a and etc.
+        <descriptor>: descriptor stores information specific to instrument. This is
+            decided by each instrument. For L0, "raw" is used.
+        <startdate>: startdate is the earliest date in the data. Format - YYYYMMDD
+        <enddate>: Some instrument and some data level requires to store date range.
+            If there is no end date, then startdate will be used as enddate as well.
+            Format - YYYYMMDD.
+        <version>: This stores software version and data version. Version format is
+            vxx-xx.
+
+        Parameters
+        ----------
+        filename : str | Path
+            Science data filename or file path.
+        data_dir : Path, optional
+            Optional higher directory level for the data, by default None
```

Comment:
What do we mean by this variable? I think we only want to go through `imap_data_access.config["DATA_DIR"]` rather than have another option as well for multiple nestings.

Reply:
Mostly I wanted to move the work out of io.py into here, it made sense to add it as a keyword argument instead of getting it from the data directory just to have more flexibility for people importing the class and using it directly rather than using it through imap_data_access io. The imap_data_access methods pass in the data dir config to the class, but in imap_processing or something I think it makes sense to keep it only in the class rather than reading it from an environment variable.

Reply:
I can remove this if preferred, I thought it was a little cleaner than handling it in `io.py` but I can go back to handling it there if this is confusing

Reply:
I was torn as well allowing an extra argument in the `io.py` routines, so I think we're thinking the same thing here :) and there is not a right/wrong answer.

My reason for not adding it was to try and start locking down potential miscues and only make one place to change things, which would be the `config` dictionary and if you want to update that it could be through the command line, environment variables, or via attribute updating within the package.

From your example in the README:
```python
# Use a data directory
imap_data_access.config["DATA_DIR"] = Path("/data/personal")
science_file = imap_data_access.ScienceFilePath("imap_swe_l0_sci_20240105_20240105_v00-05.pkts")

# Filepath = /data/personal/imap/swe/l0/2024/01/imap_swe_l0_sci_20240105_20240105_v00-05.pkts
filepath = science_file.construct_file_path()

# or have the using routines construct it themselves?
filepath = imap_data_access.config["DATA_DIR"] / science_file.construct_file_path()
```

The two options above would return either the fully qualified path directly, or have the user construct it themselves if it kept returning the current relative path. My preference would probably be the second version, and removing the `data_dir` argument option and only allowing filename.